File: leetcode/Remove Duplicates from Sorted Array.cpp
#define F(i, a, b) for(int i = a; i <= b; i++)
#include <vector>
#include <unordered_map>

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void swap( int* A, int& a, int& b){
        if( a == b ){
            return ;
        }else{
            int tmp = A[a];
            A[a] = A[b];
            A[b] = tmp;
        }
    }

    int removeDuplicates(int A[], int n) {
        if( n < 2 ){
            return n;
        }

        int tmp = A[0];
        int len = n;
        int index = 1;
        F(i, 1, n-1){
            if( A[i] == tmp ){
                len--;
                index--;
            }else{
                tmp = A[i];
                if( index < i ){
                    swap(A, i, index);
                }
            }
            index++;
        }

        return len--;
    }
};

File: LeetCode-Java/Remove Duplicates from Sorted Array.java
public class Solution {
	public int removeDuplicates(int[] A) {
		int len = A.length;
		if( len < 1 ){
			return len;
		}
		if( len < 2 ){
			return 1;
		}
		int index = 1;
		int processing = A[0];
		for( int i = 1; i < A.length; i++ ){
			if( A[i] == processing ){
				len--;
			}else{
				processing = A[i];
				if( index < i ){
					A[index] = A[i];
				}
				index++;
			}
		}
		return len;
	}
}
